File: C/swapPairs.c
/*
给你一个链表，两两交换其中相邻的节点，并返回交换后链表的头节点。
你必须在不修改节点内部的值的情况下完成本题（即，只能进行节点交换）。

输入：head = [1,2,3,4]
输出：[2,1,4,3]
输入：head = []
输出：[] 
*/
//  Definition for singly-linked list.
#include<stdio.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* swapPairs(struct ListNode* head){
    // 空
    if (head == NULL || head->next==NULL)
        return head;
    // 新链表
    struct ListNode* p = head->next;
    // 第二个节点指向后面的节点，同时后面的节点又实行一次递归
    // 两两递归一次
    head->next= swapPairs(p->next);
    // 然后将每次递归的的第二个节点指向前一个节点
    p->next = head;
    // 返回链表即可
    return p;
}

File: Java/src/leetcode24/swapPairs.java
package leetcode24;

//Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
class Solution {
   /* public ListNode swapPairs(ListNode head) {
        if (head==null || head.next==null){
            return head;
        }
        ListNode two = head.next;
        ListNode three = head.next.next;
        two.next = head;
        head.next = swapPairs(three);
        return two;
    }*/
/*    public ListNode swapPairs(ListNode head) {
        //如果当前节点和当前节点的下一节点是空，则直接返回当前节点
        if (head==null || head.next==null)
            return head;
        //定义当前节点的下一节点
        ListNode nextNode = head.next;
        //让当前节点指向下一节点的后续的节点，同时完成后面节点的递归指向
        head.next = swapPairs(nextNode.next);
        //当头结点指向“第三个节点”后 之前的第二节点就可以直接指向头结点（现在的第二节点）
        nextNode.next=head;
        return nextNode;
}*/

    public ListNode swapPairs(ListNode head){
         //如果当前节点和当前节点的下一节点是空，则直接返回当前节点
        if (head==null || head.next==null)
            return head;
        //定义当前节点的下一节点
        ListNode nextNode = head.next;
        //让当前节点指向下一节点的后续的节点，同时完成后面节点的递归指向
        head.next = swapPairs(nextNode.next);
        //当头结点指向“第三个节点”后 之前的第二节点就可以直接指向头结点（现在的第二节点）
        nextNode.next=head;
        return nextNode;
    }

}